### PR TITLE
Aurn refactored and split into extractor and post_processor.

### PR DESCRIPTION
**Pull request by AnnAnnFryingPan**
Once this is OK, we can (in no partic order)
a) finish writing doc-strings etc
b) final check of readMe
c) Doug check aurn_post_processor and met_post_processor to look at further re-use of code (sharing of functions etc) 
      E.G. station_listing() in [Met|Aurn]PostProcessor classes?
d) Write tests.
e) Ask Manuele to test

### CODE REVIEW

File: MEDMI_Data_Preprocessing/Met_Temperature_RH_Pressure_data_processing.py [hunk 1]
```diff
@@ -1,3 +1,23 @@
...
+        b) removing stations identified as unwanted
+            i) station 117 is on top of a mountain in the Cairngorms - RH readings are suspect,
+                    and as it is unlikely to be useful comparison with participant data, we will remove it
+        c) find and remove the synoptic spot readings (these are single readings per day - so we will
+                     identify all such single readings and remove them, if they are synoptic spot readings or not)
+        (Points (a) and (c) based on pers. comms. with Martyn Sunter, Met Office, July 2020.
+         Point (b) based on data exploration by authors.)
+    2)
+
+@author: mbessdl2
```
Comment (douglowe) [inline MEDMI_Data_Preprocessing/Met_Temperature_RH_Pressure_data_processing.py:18]: we should change the author attribution to us both (not just use my UoM username)

File: environmental_data_modules/aurn_extractor.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+            # tidy data, adding site column and removing excess columns
+            full_hourly_dataframe = self.tidy_hourly_data(full_hourly_dataframe, site)
+
+            # return the full hourly dataset
+            final_dataframe = final_dataframe.append(full_hourly_dataframe)
+
+            # Todo Doug: should this be removed
+            # postprocessing the data set, to get daily data
+            # final_dataframe = final_dataframe.append(postprocess_data(full_hourly_dataframe,site))
```
Comment (douglowe) [inline environmental_data_modules/aurn_extractor.py:139]: I think we should remove this commented out code, because the post processing should be separate from the downloading and preparing of the hourly data

File: environmental_data_modules/aurn_extractor.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+            if save_to_csv is True:
+                # Now save the full hourly dataframe as a .csv file
+                full_hourly_dataframe.to_csv(os.path.join(self.out_dir, site + '.csv'), index=False, header=True)
+
+        return final_dataframe
+
+        # functions for downloading data
```
Comment (douglowe) [inline environmental_data_modules/aurn_extractor.py:147]: I think this comment might be an orphan which needs removing?

File: environmental_data_modules/aurn_extractor.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        # now create list of years to process
+        years_process = []
+
+        for year in years:
+            if (year <= latest_year and year >= earliest_year):
+                years_process.append(year)
+
+        return years_process
+
+    def download_and_open_datafiles(self, site, station_name, years):
```
Comment (douglowe) [inline environmental_data_modules/aurn_extractor.py:176]: We have a number of string constructors in this routine that should be converted to "".format()

File: environmental_data_modules/aurn_extractor.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+                filename_path = os.path.join(self.out_dir, downloaded_file)
+                if os.path.exists(filename_path):
+                    print("\t\tData file already exists, will use this")
+                else:
+                    print("\t\tDownloading data file for ", station_name, " in ", str(year))
+                    wget.download(download_url, out=str(self.out_dir))
+
+                # Read the RData file into a Pandas dataframe
+                downloaded_data = pyreadr.read_r(str(filename_path))
+                # Drop non-required fields
```
Comment (douglowe) [inline environmental_data_modules/aurn_extractor.py:196]: this is an orphaned comment that should be removed

File: environmental_data_modules/aurn_extractor.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    def tidy_hourly_data(self, hourly_dataframe, site):
+
+        columns_of_interest = ['datetime', 'O3', 'NO2', 'SO2', 'NOXasNO2', 'PM2.5', 'PM10']
+        ds_columns = hourly_dataframe.columns
+
+        # retain the data we are interested in (as not all datasets have all variables)
+        columns_to_retain = set(columns_of_interest) & set(ds_columns)
+        working_dataframe = hourly_dataframe[columns_to_retain]
+
+        working_dataframe = working_dataframe.rename(columns={'datetime': 'Date'})
```
Comment (douglowe) [inline environmental_data_modules/aurn_extractor.py:224]: Perhaps we should change the code in lines 209 and 210 (of load_sort_data) so that we set the column name to Date from the get-go.

File: environmental_data_modules/dates_processor.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    def get_available_end():
+        """ Static method that gets available end date
+
+            Returns:
+                A single date as datetime: the last available date
+        """
+        cur_year = datetime.now().year
+        # Todo: Doug, when is best available end date to set, based on now() ? If it depends on Met or Aurn, we'll need
+        #  to sub-type this class
+        #  Currently set to end of previous year.
```
Comment (douglowe) [inline environmental_data_modules/dates_processor.py:88]: AURN data will be more up-to-date than the Met data I think. But, neither will be completely up-to-date.

Question is - how does the source repository react if we give an end date after their end date? If it copes fine then we can use today as the end date for our work here. Otherwise we need to work out a way of checking with the source repository to see what their end date is.

File: environmental_data_modules/dates_processor.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            raise ValueError('End date is greater than maximum date available.')
+        self.__end = date_end
+
+
+class DateYearsProcessor(DatesProcessor):
+    """
+        Class used for processing list of years
+    """
+
+    START_YEAR = 2016
```
Comment (douglowe) [inline environmental_data_modules/dates_processor.py]: We could probably allow an earlier start year than 2016.

Comment (douglowe) [reply environmental_data_modules/dates_processor.py]: The Pollen data is our limiting factor here - that seems to be only available from 2010 onwards.

Both the met and the AURN data are available before this (from mid-1970's for the AURN data, and earlier for the met data).

I think we should give a start date of 2010 - and add a check to see if the user specifies an earlier year than this (and if they do, warn them that pollen data wont be available for that period).

File: environmental_data_modules/dates_processor.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            raise ValueError('End date is greater than maximum date available.')
+        self.__end = date_end
+
+
+class DateYearsProcessor(DatesProcessor):
+    """
+        Class used for processing list of years
+    """
+
+    AVAILABLE_START_YEAR = 2016
```
Comment (douglowe) [inline environmental_data_modules/dates_processor.py:137]: I think we should have an earlier available start year than 2016.

Pollen data is the limiting factor here - as that is only available from 2010 onwards. Some met and AURN data is available from around 1970's onwards. I'd change the start year to 2010 - maybe with a warning for earlier years that there will be no pollen data.

File: environmental_data_modules/met_post_processor.py [hunk 10]
```diff
@@ -0,0 +1,655 @@
...
+            data_counts = temp_groups.count()
+            self.print_data_count_stats(data_counts)
+
+        return met_data_reduced
+
+
+    #%% function for getting two lists of stations, one for required site, one for reference sites
+
+    def station_listing(self, met_extracted_data, var_string):
+        #Todo Doug: can this be merged with station_listing() in aurn_post_processor (and put in to post_processor)
```
Comment (douglowe) [inline environmental_data_modules/met_post_processor.py:345]: It should be merged with that version of station_listing. Will need to read through and work out which to keep - this can be done later (once this PR has been sorted).

**Review by douglowe — APPROVED** [review]
It mostly looks okay to me (see a few, specific, comments for small changes).

I think we should merge this PR, and then I will go through the code properly (e.g. station_listing functions) and see what streamlining I think can be done.

File: environmental_data_modules/dates_processor.py (hunk 9 above)
Comment (douglowe) [reply environmental_data_modules/dates_processor.py:137]: Ahh - I see that I commented on the START_YEAR too - was sure I had.

Perhaps we need to merge these two, or refer from one to the other, when giving these default values?

File: MEDMI_Data_Preprocessing/Met_Temperature_RH_Pressure_data_processing.py (hunk 1 above)
Comment (AnnAnnFryingPan) [reply MEDMI_Data_Preprocessing/Met_Temperature_RH_Pressure_data_processing.py:18]: Have added similar doc-strings to other two high level scripts too.

File: environmental_data_modules/aurn_extractor.py (hunk 4 above)
Comment (AnnAnnFryingPan) [reply environmental_data_modules/aurn_extractor.py:176]: I've added a constant for the download R-data URL too, as it was hard coded in the middle of loads of code.

Comment (AnnAnnFryingPan) [reply environmental_data_modules/aurn_extractor.py:176]: (+ tested it)

File: environmental_data_modules/aurn_extractor.py (hunk 6 above)
Comment (AnnAnnFryingPan) [reply environmental_data_modules/aurn_extractor.py:224]: This also had hard coded values in so have added a constant, and tidied up the list of constants in aurn_module

File: environmental_data_modules/dates_processor.py (hunk 7 above)
Comment (AnnAnnFryingPan) [reply environmental_data_modules/dates_processor.py:88]: For end date, both extractors seem fine with current year, just goes up to today, so will allow current year.

File: environmental_data_modules/dates_processor.py (hunk 8 above)
Comment (AnnAnnFryingPan) [reply environmental_data_modules/dates_processor.py]: This is the sort of thing we'll be checking for in the tests, e.g. bad and edge cases of dates, years, locations etc

Comment (AnnAnnFryingPan) [reply environmental_data_modules/dates_processor.py]: I've set start dates to 2010 or (2010, 01, 01, 00)

File: environmental_data_modules/dates_processor.py (hunk 9 above)
Comment (AnnAnnFryingPan) [reply environmental_data_modules/dates_processor.py:137]: It's my fault, I committed a variable name change, during an open PR.
